### ISSUE TITLE
solr search to trigger an unexpected segment read operation

### ISSUE THREAD

**Issue by Puig** [basho/merge_index#31]
I keep seeing the following crash on the same node of my 5-node-cluster

```
{badarg,[{erlang,binary_to_term,[<<30767 bytes>>],[]},{mi_segment,read_seg_entry,1,[{file,"src/mi_segment.erl"},{line,419}]},{mi_segment,iterate_by_term_values,3,[{file,"src/mi_segment.erl"},{line,270}]},{mi_server,'-group_iterator/2-fun-1-'... 
```

That triggers the following error message too:

```
2014-05-06 23:55:31 =ERROR REPORT====
webmachine error: path="/solr/select"
{error,{error,{case_clause,timeout},[{riak_search_client,search_doc,8,[{file,"src/riak_search_client.erl"},{line,148}]},{riak_search_utils,run_query,7,[{file,"src/riak_search_utils.erl"},{line,283}]},{riak_solr_searcher_wm,to_json,2,[{file,"src/riak_solr_searcher_wm.erl"},{line,120}]},{webmachine_resource,resource_call,3,[{file,"src/webmachine_resource.erl"},{line,183}]},{webmachine_resource,do,3,[{file,"src/webmachine_resource.erl"},{line,141}]},{webmachine_decision_core,resource_call,1,[{file,"src/webmachine_decision_core.erl"},{line,48}]},{webmachine_decision_core,decision,1,[{file,"src/webmachine_decision_core.erl"},{line,546}]},{webmachine_decision_core,handle_request,2,[{file,"src/webmachine_decision_core.erl"},{line,33}]}]}}
```

I've teared the node down and the search query doesn't time out any more, but I want to know what the root of this issue is. I am running riak 1.4.0rc1.

**Comment by Puig** [comment]
I've tried to cure the issue by adding a new riak node to the ring wanting to tear that one with the issue down but now it seems that the issue has been propagated. Not exactly the same parent call (mi_segment:read_seg_entry/1) but mi_segment:iterate_all_bytes/2.

```
2014-05-07 16:55:51 =ERROR REPORT====
Error in process <0.2260.2541> on node 'cloudshare-filestore@filestore-01.cloudshare-staging.ooze.lindenlab.com' with exit value: {badarg,[{erlang,binary_to_term,[<<30619 bytes>>],[]},{mi_segment,iterate_all_bytes,2,[{file,"src/mi_segment.erl"},{line,167}]},{mi_server,'-group_iterator/2-fun-1-',2,[{file,"src/mi_server.erl"},{line,725}]},{mi_server,iterate2,5,[{file,"src... 

2014-05-07 16:55:51 =ERROR REPORT====
Error in process <0.2261.2541> on node 'cloudshare-filestore@filestore-01.cloudshare-staging.ooze.lindenlab.com' with exit value: {badarg,[{erlang,binary_to_term,[<<18915 bytes>>],[]},{mi_segment,iterate_all_bytes,2,[{file,"src/mi_segment.erl"},{line,167}]},{mi_server,'-group_iterator/2-fun-1-',2,[{file,"src/mi_server.erl"},{line,725}]},{mi_server,'-group_iterator/2-fun-0-'... 
```

**Comment by bsparrow435** [comment]
@Puig this looks like you are trying to index a very large term. @rzezeski can confirm but i believe this is expected when using a term greater than 32kb..

**Comment by Puig** [comment]
I would appreciate a confirmation. Also, I would like to know if there's an option to avoid this issue if same terms are being used.

**Comment by jonmeredith** [comment]
Terms > 32kb cause problems.  Although search is being replaced by our new solr backed implementation in 2.0 (yokozuna) we've fixed up the original search for 2.0 with this patch to prevent problems for people still migrating - https://github.com/basho/riak_search/pull/161 .  

The patch should apply fairly cleanly if you want to build from source.

**Comment by rzezeski** [comment]
@Puig I'm fairly sure there is a known corruption issues in 1.4.0rc1.  You should not use release candidates in production.  I recommend shutting down your nodes one by one, deleting all merge index files on any machines which is showing the above errors, and then upgrading to 1.4.8.  After all nodes have been upgraded you'll then need to re-index all your KV data.

Also, as @jonmeredith said, large terms will cause overflow issues which lead to corruption.  Be careful not to index binary data or large strings with no spaces in them.
